codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -0,0 +1,642 @@
+                                )
+                            }
+                        }
+                    }
+                    val status =
+                        variantShape.getTrait<HttpErrorTrait>()?.let { trait -> trait.code }
+                            ?: errorTrait.defaultHttpStatusCode
+                    rust("response = response.status($status);")
+                }
+            }
+            rust(
+                """
+                    #TKind::Unhandled(_) => {
+                        object.key(${"code".dq()}).string(${"Unhandled".dq()});
+                        response = response.status(500);
+                    }
+                """.trimIndent(),
+                errorSymbol
+            )
+        }
+        rust("object.finish();")
+        rustTemplate(
+            """
+                response.body(#{Bytes}::from(out))
+                    .map_err(#{Error}::BuildResponse)
+            """.trimIndent(),
+            *codegenScope
+        )
+    }
+
+    private fun RustWriter.serverRenderShapeResponseSerializer(
+        operationShape: OperationShape,
+        bindings: List<HttpBindingDescriptor>,
+    ) {
+        val structuredDataSerializer = protocol.structuredDataSerializer(operationShape)
+        structuredDataSerializer.serverOutputSerializer(operationShape).also { serializer ->
+            rust(
+                "let payload = #T(output)?;",
+                serializer
+            )
+        }
+        // avoid non-usage warnings for response
+        Attribute.AllowUnusedMut.render(this)
+        rustTemplate("let mut response = #{Http}::Response::builder();", *codegenScope)
+        for (binding in bindings) {
+            val serializedValue = serverRenderBindingSerializer(binding, operationShape)
+            if (serializedValue != null) {
+                serializedValue(this)
+            }
+        }
+        rustTemplate(
+            """
+                response.body(#{Bytes}::from(payload))
+                    .map_err(#{Error}::BuildResponse)?
+            """.trimIndent(),
+            *codegenScope,
+        )
+    }
+
+    private fun serverRenderBindingSerializer(
+        binding: HttpBindingDescriptor,
+        operationShape: OperationShape,
+    ): Writable? {
+        val member = binding.member
+        return when (binding.location) {
+            HttpLocation.HEADER, HttpLocation.PREFIX_HEADERS, HttpLocation.PAYLOAD -> {
+                logger.warning(
+                    "[rust-server-codegen] $operationShape: response serialization does not currently support ${binding.location} bindings"
+                )
+                null
+            }
+            HttpLocation.DOCUMENT -> {
+                // document is handled separately
+                null
+            }
+            HttpLocation.RESPONSE_CODE -> writable {
+                rustTemplate(
+                    """
+                        let status = output.${member.memberName.toLowerCase()}
```

Comment:
you need to use `symbolProvider.toMemberName(member)`—using shape fields directly won't work for fields that get escaped

Reply:
(It will also handle the lower casing for you), lowercase is also not right, it's snake case, btw.

Reply:
Oh, thanks. I spent a lot on this piece and it never felt right and now I see why :)